File: main/src/com/seovic/core/updater/PropertyUpdater.java
/*
 * Copyright 2009 Aleksandar Seovic
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.seovic.core.updater;


import com.seovic.core.Updater;

import com.tangosol.io.pof.PofReader;
import com.tangosol.io.pof.PofWriter;
import com.tangosol.io.pof.PortableObject;

import java.lang.reflect.Method;

import java.io.IOException;
import java.io.Serializable;


/**
 * Simple imlementation of {@link Updater} that updates single property of a
 * target value using introspection.
 *
 * @author Aleksandar Seovic  2009.06.17
 */
@SuppressWarnings("unchecked")
public class PropertyUpdater
        implements Updater, Serializable, PortableObject
    {
    // ---- constructors ----------------------------------------------------

    /**
     * Construct a <tt>BeanExtractor</tt> instance.
     *
     * @param propertyName the name of the property to extract, as defined by
     *                     the JavaBean specification
     */
    public PropertyUpdater(String propertyName)
        {
        m_propertyName = propertyName;
        }


    // ---- Updater implementation ------------------------------------------

    /**
     * {@inheritDoc}
     */
    public void update(Object target, Object value)
        {
        if (target == null)
            {
            throw new IllegalArgumentException("Updater target cannot be null");
            }

        Class targetClass = target.getClass();
        try
            {
            Method method = m_propertyMutator;
            if (method == null
                || method.getDeclaringClass() != targetClass)
                {
                m_propertyMutator = method = findWriteMethod(m_propertyName,
                        target.getClass(), value == null ? Object.class : value.getClass());
                }
            method.invoke(target, value);
            }
        catch (NullPointerException e)
            {
            throw new RuntimeException("Writeable property " + m_propertyName +
                                       " does not exist in the class "
                                       + targetClass);
            }
        catch (Exception e)
            {
            throw new RuntimeException(e);
            }
        }


    // ---- helper methods --------------------------------------------------

    /**
     * Attempt to find a write method for the specified property name.
     * <p/>
     * This method attempts to find a write method by prepending 'set' prefix to
     * the specified property name.
     *
     * @param propertyName property name
     * @param cls          class containing the property
     * @param propertyType property type
     *
     * @return write method for the property, or <tt>null</tt> if the method
     *         cannot be found
     */
    protected Method findWriteMethod(String propertyName, Class cls,
                                     Class propertyType)
        {
        String name = "set"
                      + Character.toUpperCase(propertyName.charAt(0))
                      + propertyName.substring(1);

        try
            {
            return cls.getMethod(name, propertyType);
            }
        catch (NoSuchMethodException ignore)
            {
            }

        return null;
        }


    // ---- PortableObject implementation -----------------------------------

    /**
     * Deserialize this object from a POF stream.
     *
     * @param reader  POF reader to use
     *
     * @throws IOException  if an error occurs during deserialization
     */
    public void readExternal(PofReader reader)
            throws IOException
        {
        m_propertyName = reader.readString(0);
        }

    /**
     * Serialize this object into a POF stream.
     *
     * @param writer  POF writer to use
     *
     * @throws IOException  if an error occurs during serialization
     */
    public void writeExternal(PofWriter writer)
            throws IOException
        {
        writer.writeString(0, m_propertyName);
        }


    // ---- Object methods --------------------------------------------------

    /**
     * Test objects for equality.
     *
     * @param o  object to compare this object with
     *
     * @return <tt>true</tt> if the specified object is equal to this object
     *         <tt>false</tt> otherwise
     */
    @Override
    public boolean equals(Object o)
        {
        if (this == o)
            {
            return true;
            }
        if (o == null || getClass() != o.getClass())
            {
            return false;
            }

        PropertyUpdater that = (PropertyUpdater) o;
        return m_propertyName.equals(that.m_propertyName);
        }

    /**
     * Return hash code for this object.
     *
     * @return this object's hash code
     */
    @Override
    public int hashCode()
        {
        return m_propertyName.hashCode();
        }

    /**
     * Return string representation of this object.
     *
     * @return string representation of this object
     */
    @Override
    public String toString()
        {
        return "PropertyUpdater{" +
               "propertyName='" + m_propertyName + '\'' +
               '}';
        }

    // ---- data members ----------------------------------------------------

    /**
     * Property name.
     */
    private String m_propertyName;


    /**
     * Property accessor.
     */
    private transient Method m_propertyMutator;
    }

File: main.net/src/Coherence.Tools/Core/Updater/PropertyUpdater.cs
using System;
using System.Reflection;
using Tangosol.IO.Pof;

namespace Seovic.Core.Updater
{
    /// <summary>
    /// Simple imlementation of <see cref="IUpdater"/> that updates single 
    /// property of a target object using reflection.
    /// </summary>
    /// <author>Ivan Cikic  2010.02.05</author>
    /// <author>Aleksandar Seovic  2010.02.05</author>
    [Serializable]
    public class PropertyUpdater : IUpdater, IPortableObject
    {
        #region Constructors

        /// <summary>
        /// Deserialization constructor (for internal use only).
        /// </summary>
        public PropertyUpdater()
        {
        }

        /// <summary>
        /// Construct <code>PropertyUpdater</code> instance.
        /// </summary>
        /// <param name="propertyName">
        /// The name of the property to update.
        /// </param>
        public PropertyUpdater(string propertyName)
        {
            m_propertyName = propertyName;
        }

        #endregion

        #region IUpdater implementation

        public void Update(object target, object value)
        {
            if (target == null)
            {
                throw new ArgumentException("Updater target cannot be null");
            }

            Type targetType = target.GetType();
            try
            {
                PropertyInfo property = m_property;
                if (property == null || property.DeclaringType != targetType)
                {
                    m_property = property 
                        = targetType.GetProperty(m_propertyName, BINDING_FLAGS);
                }
                if (property.CanWrite)
                {
                    property.SetValue(target, value, null);
                }
                
            }
            catch (NullReferenceException)
            {
                throw new Exception("Writeable property " + m_propertyName +
                                       " does not exist in the class "
                                       + targetType);
            }
        }

        #endregion

        #region IPortableObject

        public void ReadExternal(IPofReader reader)
        {
            m_propertyName = reader.ReadString(0);
        }

        public void WriteExternal(IPofWriter writer)
        {
            writer.WriteString(0, m_propertyName);
        }

        #endregion

        #region Object methods

        public bool Equals(PropertyUpdater other)
        {
            if (ReferenceEquals(null, other)) return false;
            if (ReferenceEquals(this, other)) return true;
            return Equals(other.m_propertyName, m_propertyName);
        }

        public override bool Equals(object obj)
        {
            if (ReferenceEquals(null, obj)) return false;
            if (ReferenceEquals(this, obj)) return true;
            if (obj.GetType() != typeof (PropertyUpdater)) return false;
            return Equals((PropertyUpdater) obj);
        }

        public override int GetHashCode()
        {
            return m_propertyName.GetHashCode();
        }

        public override string ToString()
        {
            return "PropertyUpdater{" +
               "propertyName='" + m_propertyName + '\'' +
               '}';
        }

        #endregion

        #region Data members

        private const BindingFlags BINDING_FLAGS =
            BindingFlags.Public | BindingFlags.Instance | BindingFlags.IgnoreCase;

        /// <summary>
        /// Property name.
        /// </summary>
        private String m_propertyName;

        /// <summary>
        /// Property info.
        /// </summary>
        [NonSerialized]
        private PropertyInfo m_property;

        #endregion
    }
}
